elastic/kibana
# Improve performance of the Logstash Pipeline Viewer

_This PR targets a feature branch instead of `master`. Once this PR is merged, a follow up PR will be made to the feature branch to update tests. The idea here is to keep the size of each PR small. Finally, a PR will be created from the feature branch to `master`._

Resolves #27513.

## Before this PR
The Logstash Pipeline Viewer UI would make a single Kibana API call to fetch all the information necessary to render the Logstash pipeline. This included information necessary to render the detail drawer that opens up when a user clicks on an individual vertex in the pipeline.

Naturally, this single API call fetched _a lot_ of data, not just from the Kibana server but also, in turn, from Elasticsearch as well. The "pro" of this approach was that the user would see instantaneous results if they clicked on a vertex in a pipeline and opened the detail drawer for that vertex. The "cons" were the amount of computation Elasticsearch had to perform and the amount of data being transferred over the wire between Elasticsearch and the Kibana server as well as between the Kibana server and the browser.

## With this PR
This PR makes the Kibana API call to fetch data necessary for **initially** rendering the pipeline — that is, with the detail drawer closed — much lighter. When the user clicks on a vertex in a pipeline, a second API call is then made to fetch data necessary for the detail drawer.

Concretely, this PR makes the following changes:

* [x] For each metric of each vertex in the pipeline viewer, instead of returning a timeseries in the Kibana API response, we now only return the _latest_ data point. In other words, instead of returning `[ [t1, v1], [t2, v2], ... [tn, vn] ]` we now only return `vn`. This dramatically cuts down the number of aggregation buckets that Elasticsearch needs to generate.

* [x] For each vertex in the pipeline viewer, instead of returning up to 9 metrics in the Kibana API response we now only return 4. These are the 4 that are actually needed for the initial rendering of the pipeline (i.e. with the detail drawer closed). The 5 metrics that are no longer returned are `events_in`, `events_out`, `duration_in_millis`, `queue_push_duration_in_millis`, and `queue_push_duration_in_millis_per_event`.  

* [x] Since we no longer return `queue_push_duration_in_millis`, and `queue_push_duration_in_millis_per_event` from the Kibana API response, we also no longer compute the aggregations behind these metrics. Specifically, we no longer need the `queue_push_duration_in_millis_temp_by_ephemeral_id` and `queue_push_duration_in_millis_total` aggregations.

* [x] We no longer return the `pipeline.representation.plugins` array in the Kibana API response, since this array is not currently used in the Pipeline Viewer UI. We do this by excluding the `plugins` field from the Elasticsearch API response for the `logstash_state` document.

* [x] A new API to return timeseries metrics for a single vertex is created. This API is called whenever the detail drawer is open and showing a plugin vertex.

## Gains, by the numbers

Based on a simple, 1-input, 1-filter, and 1-output pipeline.

* Before this PR, the Elasticsearch `logstash_stats` API responses (multiple calls were made using the `composite` aggregation over the `date_histogram` aggregation) generated a total of 1228 aggregation buckets (before any `filter_path`s were applied but across all `composite` "pages"). With this PR, the single `logstash_stats` API response (note that this is just for the initial rendering of the pipeline, with the detail drawer closed) generated 12 buckets (also before any `filter_path`s were applied). That's a **99.02% reduction** in number of buckets.

* Before this PR, the Elasticsearch `logstash_stats` API responses added up to 70319 bytes. With this PR, the single `logstash_stats` API response for the same pipeline is 746 bytes. That's a **98.93% reduction** in size.

* Before this PR, the Elasticsearch `logstash_state` API response was 7718 bytes. With this PR, the API response for the same pipeline is 2328 bytes. That's a **69.83% reduction** in size.

* Before this PR the Kibana API response was 51777 bytes. With this PR, the API response for the same pipeline is 2567 bytes (again, note that this is just for the initial rendering of the pipeline, with the detail drawer closed). That's a **95.04% reduction** in size.

## Testing this PR

1. [Setup Logstash Monitoring](https://www.elastic.co/guide/en/logstash/current/configuring-logstash.html).
2. Start a Logstash pipeline, like so:

   ```
   ./bin/logstash -e 'input { generator {} } filter { drop {} } output { stdout { codec => dots } }'
   ```
3. Let the pipeline run for ~30 minutes (due to [this bug](https://github.com/elastic/kibana/issues/28153) 😞).

4. Open the Monitoring UI and visit the Logstash Pipeline Viewer for your pipeline.

5. Make sure the pipeline shows up. Check that the API response for the initial rendering of the pipeline doesn't contain any timeseries data. Then click on a vertex to open the Detail Drawer. Check that the details for that vertex show up, including sparklines rendered using timeseries data. Sample screenshots of what to expect:

   * Initial rendering

   <img width="1182" alt="Screen Shot 2019-03-18 at 11 02 45 AM" src="https://user-images.githubusercontent.com/51061/54552538-87caf980-496d-11e9-93ad-ede1eb6defa9.png">

   * With detail drawer open for `drop` filter plugin vertex

   <img width="1201" alt="Screen Shot 2019-03-18 at 11 02 57 AM" src="https://user-images.githubusercontent.com/51061/54552548-8dc0da80-496d-11e9-9e9e-8d130ed361ca.png">
